# Introduction


In this project, we will work with a dataset of used cars from eBay Kleinanzeigen, which is a classified section of the eBay website.

The dataset for this project can be retrieved from (data.world)[https://data.world/data-society/used-cars-data].

This dataset contains 50,000rows and 20 columns.

## Information about each column

Since the original dataset was scrapped from a German site, its original form was in German and has been translated to english.

Below is a list of all the columns in the dataset and their descriptions.


| **Column**       |   **Description**|
|  :----------      |    :----------    |
| dateCramwled     |    Gives information about when the ad was first |crawled. All fields value are taken from this date.|
|name             |    Name of the car|
|seller           |    Whether the seller is private or a dealer.|
|offerType        |    The type of listing.|
|price            |    The price on the ad to sell the car.|
|abtest           |    Whether the listing is included in an A/B test.|
|vehicleType      |    The vehicle Type.|
|yearOfRegistration    |The year in which the car was first registered.|
|gearbox  | The transmission type|
|powerPS | The power of the car in PS.|
|model | The car model name.|
|kilometer | How many kilometers the car has driven.|
|monthOfRegistration | The month in which the car was first registered.|
|fuelType | What type of fuel the car uses.|
|brand | The brand of the car.|
|notRepairedDamage | If the car has a damage which is not yet repaired.|
|dateCreated | The date on which the eBay listing was created.|
|nrOfPictures | The number of pictures in the ad.|
|postalCode | The postal code for the location of the vehicle.|
|lastSeenOnline | When the crawler saw this ad last online.|

# GOALS

### The goal of this project includes:
- Cleaning the dataset
- Analyse the used car listings.
- Determining the relationship between price and mileage


Let's start by importing the required libraries and reading our dataset.

In [80]:
# Import pandas and numpy

import pandas as pd
import numpy as np

# Read the dataset
autos = pd.read_csv('autos.csv', encoding='Latin-1')



Reading the first few and last few rows.

In [81]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Let's display information about the dataframe and the first 5rows

In [82]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can observe that the dataset has two major datatypes, int64 and objects, with 75% being objects.

Also, we notice that some columns have null values. There are some columns that contain dates stored as strings.

We will start by cleaning the columns names to make it easier to work with them.
We will convert the camelcase columns to snakecase.

In [83]:
autos.columns = ['date_crawled', 'name', 'seller','offer_type', 'price', 'ab_test', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code', 'last_seen']
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  num_pictures  postal_code  \
0              nein  2016-03-26 00:00:00             0        79588   
1              nein  2016-04-04 00:00:00             0        71034   
2              nein  2016-03-26 00:00:00             0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

# Data Exploration

Let's explore our data to find determine the columns that need to be cleaned.

In [84]:
autos.describe( include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

# Observation

According to the information displayed above, we can observe that columns such as : 
- seller 
- offer_type 
- ab_test
- gearbox 
- unrepaired_damage 
have values which are all or almost all the same. We can therefore get rid of some of them since they will not influence our analysis.


The num_pictures column seems odd. It shows no unique values and no value for the mean. let's inverstigate it further.



In [85]:
autos['num_pictures'].value_counts()

0    50000
Name: num_pictures, dtype: int64

All the 50000 enteries for the number of pictures has 0, we can therefore drop this column since it won't affect our analysis. 

We will also drop the seller and offer_type columns.

In [86]:
autos.drop(['seller', 'offer_type', 'num_pictures'], axis=1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       $5,000  control          bus               2004    manuell       158   
1       $8,500  control    limousine               1997  automatik       286   
2       $8,990     test    limousine               2009    manuell       102   
3       $4,350  control   kleinwagen               2007  automatik        71   
4       $1,350     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995  $24,900  control    limousine               2011  automatik       239   
49996   $1,980  control       cabrio               1996    manuell        75   
49997  $13,200     test       cabrio               2014  automatik        69   
49998  $22,900  control        kombi               2013    manuell       150   
49999   $1,250  control    limousine               1996    manuell       101   

        model   odometer  registration_month fuel_type       brand  \
0      andere  150,000km                   3       lpg     peugeot   
1         7er  150,000km                   6    benzin         bmw   
2        golf   70,000km                   7    benzin  volkswagen   
3      fortwo   70,000km                   6    benzin       smart   
4       focus  150,000km                   7    benzin        ford   
...       ...        ...                 ...       ...         ...   
49995      q5  100,000km                   1    diesel        audi   
49996   astra  150,000km                   5    benzin        opel   
49997     500    5,000km                  11    benzin        fiat   
49998      a3   40,000km                  11    diesel        audi   
49999  vectra  150,000km                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

We can see that the price has a minimum value of $0, let us investigate it further, knowing that a car can't be sold at that price.

Let's start by cleaning the price column by converting the string values to corresponding float values.

In [87]:
autos['price'] = (autos['price']
                  .str.replace('$', '')
                  .str.replace(',', '')
                  .astype(float))

print(autos['price'].head())

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64


In [88]:
print(autos['price'].value_counts())
print(autos['price'].describe())

0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
6202.0        1
18310.0       1
898.0         1
11240.0       1
789.0         1
Name: price, Length: 2357, dtype: int64
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


about 1500 cars have a price of 0 dollars. Since this number(1500) represents less than 2% of our data sample, we can get rid of the rows with prices of 0 dollar.

We can observe that the prices for some cars are above 100 Million dollars. which seems alot for a reused car. Let's have a closer look at the highest prices.



In [89]:
# Display the price count in descending order
print(autos['price'].value_counts().sort_index(ascending=False).head(20))

print('ASCENDING ORDER')
# Display the price count in ascending order
print(autos['price'].value_counts().sort_index().head(20))

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64
ASCENDING ORDER
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64


There are about 14 price listing with values above 1million dollars. There are some price listing with values below 30 dollars  and about 1500 with prices of 0 dollars

We will get rid of those with prices below 0 dollars and the rows with prices above 350000 dollars because there was a surge in price after 350000 dollars

In [90]:
autos[autos['price'].between(1, 35000)]
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

### Next, let's look at the ondometer column.

In [91]:
print(autos['odometer'].describe(include='all'))
print(autos['odometer'].value_counts())

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object
150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64


The odometer does not have any exagerated value. however, the numeric values of the odometer are stored as strings. let us clean that and rename the column to odometer_km

In [92]:
autos['odometer'] = (autos['odometer']
                    .str.replace(',', '')
                    .str.replace('km', '')
                    .astype(float))

autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

## Exploring date related columns

The date related columns include: 
   - date_crawled
   - last_seen
   - ad_created
   - registration_month
   - registration_year
   
However, what interest us are the first 3 since they contain string values. We will convert them to numeric value to fascillitate working with them.

In [93]:
print(autos[['date_crawled','ad_created','last_seen']][0:5])

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


We can observe that the have the same format, the first 10 strings representing the date followed the remaining representing the time.

We will seperate the date from the time for each column and investigate each of them.

In [94]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

From our observation, the site was crawled daily without a skip from March till April

In [95]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

The last seen date indicates when the crowler last saw the ads, meaning when the car was sold. These values are uniform from march 5th to April 4th.

We can observe a disproportionate increase in the number of last seens between April 4th to April 7th. This increase is almost 10 times the previous days.

This could be due to massive sells of cars or due to the ending period of the website crawling.

In [96]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

There is a great variety in the ads creation date. With very few created 9months before the listing and most of the created withing one month of the listing.

In [97]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can observe some unrealistic values, the minimum value is the year 1000, long before the evolution of technology, and the maximum value is 9999 which is thousands of years from now. 

Since a car can't be listed before being registered, all registered years above 2016 will be considered invalid. Also, since our dataset involves used cars, it is saved to assume that the cars have been bought and used few years before 2016 before being listed for sell on eBay.

For this reason, we will assume that a valid registration year lies between 1990 and 2016. 

Let's study this year range.

In [98]:
(~autos["registration_year"].between(1990,2016)).sum() / autos.shape[0]

0.06728

from the above, less than 6% of the car registration year falls inside the invalid year. Since this number is small, the rows with invalid years can therefore be dropped. 

In [99]:
autos = autos[autos["registration_year"].between(1990,2016)]

print(autos['registration_year'].value_counts(normalize=True))

2000    0.071919
2005    0.064650
1999    0.064328
2004    0.058689
2003    0.058474
2006    0.058067
2001    0.057960
2002    0.054314
1998    0.052599
2007    0.049404
2008    0.047839
2009    0.044987
1997    0.043486
2011    0.035037
2010    0.034244
1996    0.030963
2012    0.028369
2016    0.028219
1995    0.028154
2013    0.017283
2014    0.014281
1994    0.014152
1993    0.009542
2015    0.008556
1990    0.008470
1992    0.008384
1991    0.007634
Name: registration_year, dtype: float64


Most of the vehicles where first registered about 15years before 2016. 

# Let's study the most popular car brands and their average price.

In [100]:
# Get the brand counts 
brand_counts = autos['brand'].value_counts(normalize=True)
print(brand_counts)


volkswagen        0.211746
bmw               0.111223
opel              0.109400
mercedes_benz     0.092675
audi              0.087851
ford              0.069860
renault           0.048418
peugeot           0.030384
fiat              0.025860
seat              0.018698
skoda             0.016404
mazda             0.015482
nissan            0.015353
smart             0.014324
citroen           0.013938
toyota            0.012630
hyundai           0.010142
volvo             0.009027
mini              0.008791
mitsubishi        0.008341
sonstige_autos    0.008084
honda             0.007869
kia               0.007312
alfa_romeo        0.006390
suzuki            0.006025
porsche           0.005404
chevrolet         0.005103
chrysler          0.003667
dacia             0.002637
daihatsu          0.002595
jeep              0.002230
subaru            0.002166
land_rover        0.002016
saab              0.001565
daewoo            0.001544
jaguar            0.001544
rover             0.001394
l

volkswagen is the most popular brand accounting for almost 21% of the total brands followed by bmw. Overall, German brands are by far the most popular having more than 50% of popularity.

Since most of the brands account for less than 1% of total listing, we will focus our analysis with brands having atleast 4% of total listing.

In [101]:
# Brands waith at least 4% of total listing.

top_car_brand = brand_counts[brand_counts > 0.04].index
print(top_car_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


Let's determine the mean price for each of the top brands

In [102]:
mean_brand_price = {}
for brand in top_car_brand:
    mean_price = autos.loc[autos['brand'] == brand, 'price'].mean()
    mean_brand_price[brand] = int(mean_price)
    
    
print(mean_brand_price)

{'volkswagen': 6358, 'bmw': 8372, 'opel': 5278, 'mercedes_benz': 31582, 'audi': 9181, 'ford': 3668, 'renault': 2342}


The most expensive car brand on average is the mercedes_benz with  a price almost 12times more expensive than the least expensive brand, renault.
Majority of brands have an average lsiting price between five to ten thousand dollars. With about $8000   you can get most of the cars on the website.

# Exploring Mileage for each brand

Let's calculate the mean mileage for each popular brand, then create a dataframe that contains both the mean price and mean mileage for each brand.

In [103]:
mean_brand_mileage = {}
for brand in top_car_brand:
    mean_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    mean_brand_mileage[brand] = int(mean_mileage)

    
print(mean_brand_mileage)

{'volkswagen': 129129, 'bmw': 132565, 'opel': 129840, 'mercedes_benz': 130930, 'audi': 129090, 'ford': 125164, 'renault': 128461}


## Converting both the mean price and the mean mileage dictionary into a common dataframe object

In [104]:
# Converting mean price into a series object.
mean_prices_series = pd.Series(mean_brand_price)

# Converting mean mileage into a series object.
mean_mileage_series = pd.Series(mean_brand_mileage)

# Converting the mean price series object into a dataframe object
brand_info = pd.DataFrame(mean_prices_series, columns=['mean_price'])

# Adding the mean mileage column to our dataframe
brand_info['mean_mileage'] = mean_mileage_series
print(brand_info)

               mean_price  mean_mileage
volkswagen           6358        129129
bmw                  8372        132565
opel                 5278        129840
mercedes_benz       31582        130930
audi                 9181        129090
ford                 3668        125164
renault              2342        128461


# End of Analysis

Though the average mileage value for each brand varies less than the average price, we can observer a slight trend between average mileage and average price. Cars with smaller mileage values tend to be less expensive than those with higher mileage values.